api/src/services/forms.js

```diff
@@ -26,21 +28,48 @@ module.exports = {
     return name && doc._attachments[name];
   },
 
+  getAttachment: async (doc, name) => {
+    try {
+      return await db.medic.getAttachment(doc._id, name, { rev: doc._rev });
+    } catch (error) {
+      if (error.status === 404) {
+        logger.error(error);
+        return null;
+      }
+      throw error;
+    }
+  },
   /**
    * @return {Promise} Resolves all the documents of type 'form'
    *   with a valid xform attachment.
    */
-  getFormDocs: () => {
-    return db.medic
-      .allDocs({
-        startkey: FORM_DOC_ID_PREFIX,
-        endkey: `${FORM_DOC_ID_PREFIX}\ufff0`,
-        include_docs: true,
-      })
-      .then(response => {
-        return response.rows
-          .map(row => row.doc)
-          .filter(doc => doc && doc.type === DOC_TYPE && module.exports.getXFormAttachment(doc));
-      });
+  getFormDocs: async () => {
+    const response = await db.medic.allDocs({
+      startkey: FORM_DOC_ID_PREFIX,
+      endkey: `${FORM_DOC_ID_PREFIX}\ufff0`,
+      include_docs: true,
+    });
+
+    const docs = await Promise.all(response.rows.map(async row => {
+      const doc = row.doc;
+      const name = module.exports.getXFormAttachmentName(doc);
+      if (doc.type !== DOC_TYPE || !name) {
+        return null;
+      }
+
+      const xml = await module.exports.getAttachment(doc, name);
+      if (xml===null){
+        return null;
+      }
```

Comment:
I think this can be simplified: 
```suggestion
if (!xml) {
  return;
}
```